water/lib/Dialect/Wave/IR/WaveOps.cpp

```diff
@@ -1129,14 +1113,104 @@ LogicalResult MmaOp::verify() {
 }
 
+/// Compute the expected elements per thread for this MMA operation.
+/// Extracts threadsPerWave from ancestor operations with hardware constraints.
+/// Returns 0 if no constraints are found.
+unsigned wave::MmaOp::computeElementsPerThread() {
+  if (!getKindAttr()) {
+    return 0;
+  }
+  wave::WaveMmaSpec spec =
+      wave::WaveMmaKindAttr::getSpec(getContext(), getKind());
+
+  // Extract threads per wave from hardware constraint by walking up the
+  // ancestry.
+  mlir::Operation *op = getOperation();
+  while (op) {
+    if (auto constraints = op->getAttrOfType<mlir::ArrayAttr>(
+            wave::WaveDialect::kWaveConstraintsAttrName)) {
+      for (mlir::Attribute constraint : constraints) {
+        if (auto hardwareConstraint =
+                llvm::dyn_cast<wave::HardwareConstraintAttr>(constraint)) {
+          unsigned totalElements = spec.m * spec.n;
+          return totalElements / hardwareConstraint.getThreadsPerWave();
+        }
+      }
+    }
+    op = op->getParentOp();
+  }
+
+  // Return 0 to indicate failure if no constraints found.
+  return 0;
+}
+
+llvm::FailureOr<mlir::ChangeResult>
+wave::MmaOp::propagateElementsPerThreadForward(
+    llvm::ArrayRef<wave::ElementsPerThreadLatticeValue> operandElements,
+    llvm::MutableArrayRef<wave::ElementsPerThreadLatticeValue> resultElements,
+    llvm::raw_ostream &errs) {
+  unsigned expectedElementsPerThread = computeElementsPerThread();
+  if (expectedElementsPerThread == 0) {
+    errs << "MMA operation has no hardware constraints available";
+    return mlir::failure();
+  }
+  wave::ElementsPerThreadLatticeValue expectedResult(expectedElementsPerThread);
+  return wave::detail::checkAndPropagateElementsPerThreadFromConstant(
+      expectedResult, llvm::ArrayRef<wave::ElementsPerThreadLatticeValue>(),
+      resultElements, "computed from MMA kind", "", "result", errs);
+}
+
+llvm::FailureOr<mlir::ChangeResult>
+wave::MmaOp::propagateElementsPerThreadBackward(
+    llvm::MutableArrayRef<wave::ElementsPerThreadLatticeValue> operandElements,
+    llvm::ArrayRef<wave::ElementsPerThreadLatticeValue>,
+    llvm::raw_ostream &errs) {
+  // For MMA, the accumulator should have the same elements per thread as the
+  // result. The LHS and RHS operands may have different constraints based on
+  // their dimensions.
+  unsigned expectedElementsPerThread = computeElementsPerThread();
+  if (expectedElementsPerThread == 0) {
+    errs << "MMA operation has no hardware constraints available";
+    return mlir::failure();
+  }
+  wave::ElementsPerThreadLatticeValue expectedAccumulator(
+      expectedElementsPerThread);
+
+  unsigned accumulatorOperandNumber =
+      getAccumulatorMutable().getOperandNumber();
+
+  // Validate that LHS and RHS operands have concrete elements_per_thread
+  // values. We don't propagate to them, but we check they've been properly
+  // initialized.
+  for (unsigned i = 0; i < 2 && i < operandElements.size();
```

Comment:
```suggestion
  for (unsigned i = 0; i < min(2, operandElements.size());
```

Reply:
Also, precompute this value so it is not re-evaluated on each iteration :)